tests/integration/query_engines/test_mssql.py

```diff
@@ -78,7 +78,7 @@ def wrapper(self, *args, **kwargs):
                 # itself. But I've tried and failed to do so, partly because the
                 # `pymssql` driver we use is a compiled library and so less amenable to
                 # monkey patching.
-                self._dbapi_connection = None
+                self.invalidate()
```

Comment:
This looks completely reasonable, but its not clear why its included in this change?

Reply:
Ah yeah, sorry. It's because calling `invalidate()` causes the connection to actually get closed rather than left in the connection pool. With this in place the old retry code fails, but the new code succeeds.